apps/src/codebridge/FilePreview/InnerHTMLPreview.tsx

```diff
@@ -0,0 +1,223 @@
+            fileType = file.language;
+          }
+          const blob = new Blob([file.contents], {type: fileType});
+          const fullFileName = getFullyQualifiedFileName(
+            file.name,
+            file.folderId,
+            source.folders
+          );
+          files[fullFileName] = URL.createObjectURL(blob);
+        }
+      });
+      // Handle HTML files. We replace src links to non-html files with blob URLs.
+      // We update links to other files with a click handler that will post a message to us
+      // to change the file.
+      const htmlFiles = Object.values(source.files).filter(
+        file => file.language === 'html'
+      );
+      htmlFiles.forEach(file => {
+        const parser = new DOMParser();
+        const doc = parser.parseFromString(file.contents, 'text/html');
+        // Remove any existing CSP meta tags, as we need to set our own.
+        const existingCspTags = doc.querySelectorAll(
+          'meta[http-equiv="Content-Security-Policy"]'
+        );
+        existingCspTags.forEach(tag => tag.remove());
+
+        const metaTag = doc.createElement('meta');
+        metaTag.setAttribute('http-equiv', 'Content-Security-Policy');
+        // TODO: Improve the list of allowed origins.
+        // https://codedotorg.atlassian.net/browse/CT-579
+        metaTag.setAttribute(
+          'content',
+          "connect-src 'self' http://numbersapi.com"
+        );
+
+        const head = doc.querySelector('head');
+        if (head) {
+          head.appendChild(metaTag);
+        }
+        const links = doc.querySelectorAll(
+          'link[rel="stylesheet"], script[src]'
+        );
+        links.forEach(link => {
+          const src = link.getAttribute('src') || link.getAttribute('href');
+          if (src && files[src]) {
+            const blobUrl = files[src];
+            if (link.tagName.toLowerCase() === 'link') {
+              link.setAttribute('href', blobUrl);
+            } else {
+              link.setAttribute('src', blobUrl);
+            }
+          }
+        });
+        const fileLinks: NodeListOf<HTMLAnchorElement> =
+          doc.querySelectorAll('a[href]');
+        fileLinks.forEach(link => {
+          const href = link.getAttribute('href');
+          if (href?.endsWith('.html')) {
+            link.setAttribute(
+              'onclick',
+              `event.preventDefault();
+              window.parent.postMessage({type: '${IframeMessageType.CHANGE_FILE_HREF}', fileName: '${href}'}, '${location.origin}');
+              return false;
+            `
+            );
+          }
+        });
+        const updatedContents = doc.documentElement.outerHTML;
+        const blob = new Blob([updatedContents], {type: 'text/html'});
+        const fullFileName = getFullyQualifiedFileName(
+          file.name,
+          file.folderId,
+          source.folders
+        );
+        files[fullFileName] = URL.createObjectURL(blob);
```

Comment:
Confirming my understanding: every time source is updated, is a new [object URL](https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL_static) for the blob created. Should we clean up these old object URLs with `revokeObjectUrl`?

Reply:
good idea!